tripperware/rate_limit.go

```diff
@@ -0,0 +1,55 @@
+package tripperware
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/gol4ng/httpware/v2"
+	"github.com/gol4ng/httpware/v2/rate_limit"
+)
+
+const RequestLimitReachedErr = "request limit reached"
+
+type RateLimitOptions struct {
+	retry bool
+	sleepDuration time.Duration
+	limitRetry int
+	errorCallback func(err error, req *http.Request) error
+}
+
+type RateLimitOption func(*RateLimitOptions)
+
+func NopErrorCallback() func(err error, req *http.Request) error {
+	return func(err error, req *http.Request) error {
+		return err
+	}
+}
+
+func WithErrorCallback(callback func(err error, req *http.Request) error) RateLimitOption {
+	return func(args *RateLimitOptions) {
+		args.errorCallback = callback
+	}
+}
+
+func RateLimit(rateLimiter rate_limit.RateLimiter, options ...RateLimitOption) httpware.Tripperware {
+	args := &RateLimitOptions{
```

Comment:
create a `func NewRateLimitOptions(options ...RateLimitOption)`
and create `func (o * RateLimitOptions)apply(options ...RateLimitOption)`